notebooks/classification/Cifar10.md

```diff
@@ -0,0 +1,481 @@
+
+```python
+method_params = {
+    "naive": ("naive", False),
+    "score": ("score", False),
+    "cumulated_score": ("cumulated_score", True),
+    "random_cumulated_score": ("cumulated_score", "randomized"),
+    "top_k": ("top_k", False)
+}
+alpha = 0.1
+```
+
+```python tags=[]
+y_preds, y_pss = {}, {}
+alphas = np.arange(0.01, 1, 0.01)
+for name, (method, include_last_label) in method_params.items():
+    mapie_model = TensorflowToMapie(model)
+    mapie = MapieClassifier(estimator=mapie_model, method=method, cv="prefit", random_state=42) 
+    mapie.fit(X_calib, y_calib, image_input=True)
+    y_preds[name], y_pss[name] = mapie.predict(X_test, alpha=alphas, include_last_label=include_last_label)
+```
+
+Let's now estimate the number of null prediction sets, marginal coverages, and averaged prediction set sizes obtained with the different methods for all confidence levels and for a confidence level of 90 \%.
+
+```python
+def count_null_set(y):
+    count = 0
+    for pred in y[:, :]:
+        if np.sum(pred) == 0:
+            count += 1
+    return count
+```
+
+```python
+nulls, coverages, accuracies, sizes = {}, {}, {}, {}
+for name, (method, include_last_label) in method_params.items():
+    accuracies[name] = accuracy_score(y_true, y_preds[name])
+    nulls[name] = [
+        count_null_set(y_pss[name][:, :, i])  for i, _ in enumerate(alphas)
+    ]
+    coverages[name] = [
+        classification_coverage_score(
+            y_test, y_pss[name][:, :, i]
+        ) for i, _ in enumerate(alphas)
+    ]
+    sizes[name] = [
+        y_pss[name][:, :, i].sum(axis=1).mean() for i, _ in enumerate(alphas)
+    ]
+```
+
+```python
+coverage_90 = {method: coverage[9] for method, coverage in coverages.items()}
+null_90 = {method: null[9] for method, null in nulls.items()}
+width_90 = {method: width[9] for method, width in widths.items()}
+y_ps_90 = {method: y_ps[:, :, 9] for method, y_ps in y_pss.items()}
+```
+
+Let's now look at the marginal coverages, number of null prediction sets, and the averaged size of prediction sets for a confidence level of 90 \%. 
+
+```python
+summary_df = pd.concat(
+    [
+        pd.Series(coverage_90),
+        pd.Series(null_90),
+        pd.Series(width_90)
+    ],
+    axis=1,
+    keys=["Coverages", "Number of null sets", "Average prediction set sizes"]
+).round(3)
+```
+
+```python
+summary_df
+```
+
+As expected, the "naive" method, which directly uses the alpha value as a threshold for selecting the prediction sets, does not give guarantees on the marginal coverage since this method is not calibrated. Other methods give a marginal coverage close to the desired one, i.e. 90\%. Notice that the "cumulated_score" method, which always includes the last label whose cumulated score is above the given quantile, tends to give slightly higher marginal coverages since the prediction sets are slightly too big.
```

Comment:
Is is still true that "cumulated_score" always include the last label ? Should we play with the `include_last_label` option ?